### PR TITLE
Adds nvcc as a RawKernel backend

### PR DESCRIPTION
**Pull request by sjperkins**
Adds nvcc as a backend for RawKernel (issue https://github.com/cupy/cupy/issues/1928). The nvcc.py is a cut and paste of much of the functionality in `cupy_setup_build.py` and `install/*.py` -- I've avoided refactoring this at the moment as I'd like to invite input as to whether you think this PR is useful.

### CODE REVIEW

File: cupy/core/raw.pyx [hunk 1]
```diff
@@ -21,17 +21,21 @@ cdef class RawKernel:
         name (str): Name of the kernel function.
         options (str): Compile options passed to NVRTC. For details, see
             https://docs.nvidia.com/cuda/nvrtc/index.html#group__options.
+        backend (str): {"nvrtc", "nvcc"}. Defaults to "nvcc"
```
Comment (leofang) [inline cupy/core/raw.pyx]: The docstring should be `Defaults to "nvrtc"`, not "nvcc".

Comment (sjperkins) [reply cupy/core/raw.pyx]: fixed

**Comment by niboshi** [comment]
Thank you for the PR and I'm sorry for late review.

Does it have to rely on `distutils`? How about implementing a class with the same interface as `_NVRTCProgram` (`cupy/cuda/compiler.py`)?

**Comment by kmaehashi** [comment]
I discussed with @okuta and agreed to support NVCC in RawKernel.

`cupy_setup_build.py` and `install/*.py` has lots of clutters. Originally CuPy was using NVCC, so I think it's better to borrow some code from here, instead of using above files: https://github.com/cupy/cupy/blob/v1/cupy/cuda/compiler.py

**Comment by sjperkins** [comment]
Thanks for the suggestions. I'll try find some time to incorporate them in the next while.

**Comment by sjperkins** [comment]
I think this is now almost ready. One thing I'm still not quite sure how to handle with the nvcc route is the [_preprocess](https://github.com/cupy/cupy/pull/1941/files#diff-1e9580b27140241ef3374c8d8aab06ebR195) call. `compile_with_nvcc` directly produces a cubin (a bytes type) while `compile_with_nvrtc` produces unicode from which a bytes type cubin is produced. I haven't succeeded in plugging the output of `compile_using_nvcc` into the preprocessing code, but this could be because I'm not experienced with unicode. I've tried running a `.decode('utf-8')` on the cubin but this fails on translation.

Any suggestions on how to handle this?

File: cupy/core/carray.pxi [hunk 2]
```diff
@@ -124,7 +124,7 @@ cpdef str _get_header_source():
 
 cpdef function.Module compile_with_cache(
         str source, tuple options=(), arch=None, cachd_dir=None,
-        prepend_cupy_headers=True):
+        prepend_cupy_headers=True, backend="nvrtc"):
```
Comment (niboshi) [inline cupy/core/carray.pxi]: Please use `''` for string literals for consistency. Ditto. for other strings, including error message.

Comment (sjperkins) [reply cupy/core/carray.pxi]: done

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by sjperkins** [comment]
Will take a look at the test case failures.

**Comment by niboshi** [comment]
Those failures in `TEST=chainer-py3`/`py35` (such as `TestBias`) are irrelevant. Please ignore them. I'm sorry for inconvenience.

Those in `TEST=cupy-py???` seem relevant.

**Comment by sjperkins** [comment]
> I think this is now almost ready. One thing I'm still not quite sure how to handle with the nvcc route is the _preprocess call. compile_with_nvcc directly produces a cubin (a bytes type) while compile_with_nvrtc produces unicode from which a bytes type cubin is produced. I haven't succeeded in plugging the output of compile_using_nvcc into the preprocessing code, but this could be because I'm not experienced with unicode. I've tried running a .decode('utf-8') on the cubin but this fails on translation.

OK so the test cases are failing because of a problem that I ran into previously. It doesn't seem possible to decode the test case cubin bytes produced by [nvcc into unicode](https://github.com/cupy/cupy/pull/1941/files#diff-1e9580b27140241ef3374c8d8aab06ebR149), as there's some byte that causes a failure in translation. So the following

```python
with codecs.open(cubin_path, 'rb', encoding='utf-8') as bin_file:
    return bin_file.read()
```

produces:

```traceback
UnicodeDecodeError: 'utf8' codec can't decode byte 0xbe in position 18: invalid start byte
```

Without this conversion, it's not possible for the cubin to interact with other unicode variables in compiler.py.

Could someone with more unicode experience advise on how to handle this situation

**Comment by leofang** [comment]
I am not familiar with unicode either, but does the follow piece of code help? It describes how a unicode object is generated from the ptx:
https://github.com/cupy/cupy/blob/f85374d9ab3d404dc74e196e7791e8ed84c9ad9e/cupy/cuda/nvrtc.pyx#L111-L127

**Comment by sjperkins** [comment]
> I am not familiar with unicode either, but does the follow piece of code help? It describes how a unicode object is generated from the ptx:

Thanks for the suggestion, but unfortunately it still fails on the byte at position 18.

**Comment by sjperkins** [comment]
Having thought about this a bit more, I think the issue is that `compile_using_nvrtc` returns text PTX assembly, while `compile_using_nvcc` produces a binary cubin. I probably need to modify the nvcc command within _preprocess to just return the PTX... Let me see what I can do.

**Comment by sjperkins** [comment]
OK, got it working.

File: cupy/cuda/compiler.py [hunk 3]
```diff
@@ -89,18 +115,77 @@ def compile_using_nvrtc(source, options=(), arch=None, filename='kern.cu'):
...
+            raise cex
+
+        if code_type == 'ptx':
+            with open(result_path, 'rb') as ptx_file:
+                return ptx_file.read().decode('utf-8')
+        elif code_type == 'cubin':
+            with open(result_path, 'rb') as bin_file:
+                return bin_file.read()
+        else:
+            raise ValueError("Invalid code_type '%s'" % code_type)
```
[the remark this answers is not in the document]
Comment (sjperkins) [reply cupy/cuda/compiler.py]: fixed

**Comment by niboshi** [comment]
Thanks.
Jenkins, test this please

**Comment by niboshi** [comment]
The failure seems relevant.
Could you check again?

**Comment by sjperkins** [comment]
@niboshi I've been taking a look at the logs. Somewhat difficult to diagnose as it's failing on python2 but succeeding on python3. It also succeeds on python2 on my machine.

The error is:

> CUDADriverError: CUDA_ERROR_INVALID_SOURCE: device kernel image is invalid

This seems to be a somewhat broad range of errors. For example:

1. It may be that the LD_LIBRARY_PATH needs to be set up to point at appropriate CUDA and NVIDIA drivers on the test machine. See https://github.com/fireice-uk/xmr-stak/issues/2319#issuecomment-471463296 for instance.
2. I see there's also a [stackoverflow](https://stackoverflow.com/questions/55440885/chainer-python-cuda-error-invalid-source) question for chainer on this issue. Are there multiple GPUs on the test system, for instance?

**Comment by niboshi** [comment]
@sjperkins 
Sorry, I have no idea about the error.

>Are there multiple GPUs on the test system, for instance?

Yes, we use 8-GPU machines in the Jenkins CI.

**Comment by sjperkins** [comment]
> Yes, we use 8-GPU machines in the Jenkins CI.

Are they all the same GPU or are some different from the others?

Could you run the build again to see if the problem is intermittent or reproducible?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
>> Yes, we use 8-GPU machines in the Jenkins CI.
>
>Are they all the same GPU or are some different from the others?

Yes, they're all the same GPUs.

**Comment by sjperkins** [comment]
OK, looks like its reproducible. I'll dig into this a bit more, but this may take some time as I can't reproduce locally.

**Comment by niboshi** [comment]
Thanks. Please let me know when you need to run the test on Jenkins.

**Comment by leofang** [comment]
Not sure if this helps. Just some thoughts on what could be wrong and possible means for debugging locally:
1. According to the error message, the error happens at line 249 (`mod.load(cubin)`) in `cupy/cuda/compiler.py`. 
3. But,  in order to reach there, it means a cached `.cubin` is found. No actual compiling is done.
2. In `TestRaw`, the test `test_basic` is first run with default backend `nvrtc`. So presumably the `.cubin` is already generated prior to the failing test `test_backends`.
4. Could it be that the cubins generated by `nvrtc` and `nvcc` are incompatible? 
5. It's worth trying clearing out `~/.cupy/kernel_cache/` before testing locally.
6. Clearly the test cases do not cover compiling using `nvcc` due to cacheing. To avoid this, I suggest to let `nvrtc` and `nvcc` test different kernels (e.g. one targets `sum` and the other targets `multiply`). 
7. Also, it doesn't seem necessary to have separate tests (`test_basic` and `test_backends`). They can be combined.
8. It's still unclear to me how come only `TEST=cupy-py2` failed while others succeeded...

**Comment by sjperkins** [comment]
@leofang Thanks for having a look, good to have another eye on it. I want to have another look at getting this right

>     1. According to the error message, the error happens at line 249 (`mod.load(cubin)`) in `cupy/cuda/compiler.py`.
> 
>     2. But,  in order to reach there, it means a cached `.cubin` is found. No actual compiling is done.
> 
>     3. In `TestRaw`, the test `test_basic` is first run with default backend `nvrtc`. So presumably the `.cubin` is already generated prior to the failing test `test_backends`.

Agreed, thanks for the analysis.

>     4. Could it be that the cubins generated by `nvrtc` and `nvcc` are incompatible?

Yes, they probably are. But I think that different cubins should be generated for `nvrtc` and `nvcc` because ``backend`` is used in the cache key:

https://github.com/cupy/cupy/pull/1941/files#diff-1e9580b27140241ef3374c8d8aab06ebR218

>     5. It's worth trying clearing out `~/.cupy/kernel_cache/` before testing locally.

Yes, I should indeed be doing this. Thanks for pointing this out.
 
>     6. Clearly the test cases do not cover compiling using `nvcc` due to cacheing. To avoid this, I suggest to let `nvrtc` and `nvcc` test different kernels (e.g. one targets `sum` and the other targets `multiply`).

Unfortunately (because this explanation is appealing and would resolve this issue easily!) the nvcc compiler does get called on test_backend because ``env`` uses ``backend`` in the key. I've confirmed this locally with some print statements.

There are reloads between ``test_basic`` and ``test_backends``

>     7. Also, it doesn't seem necessary to have separate tests (`test_basic` and `test_backends`). They can be combined.

Agreed, I'll remove ``test_basic``.

> 
>     8. It's still unclear to me how come only `TEST=cupy-py2` failed while others succeeded...

**Comment by leofang** [comment]
@niboshi could you please kick off Jenkins again so that I can help investigate the issue? The old test results are not available now unfortunately. Thanks!

**Comment by niboshi** [comment]
@leofang Sure. Jenkins, test this please

**Comment by leofang** [comment]
There seems to be a lot of irrelevant errors floating around. But, the backend test fails in PY2, same as before I think.

@sjperkins could you rebase your branch against the master? There are quite a lot changes since your PR, and rebasing is likely to silence the noise. But there're some conflicts to be resolved, probably because `RawKernel` has been refurnished recently. If you give me access to write to your branch, I might be able to help.

**Comment by sjperkins** [comment]
>  But there're some conflicts to be resolved, probably because `RawKernel` has been refurnished recently. If you give me access to write to your branch, I might be able to help.


Thanks, I've added you as a collaborator to the repo.

**Comment by sjperkins** [comment]
> @sjperkins are you sure you added the right ID to the right repo? I don't see any invite nor can I push to your branch.
> 
> @niboshi we just dropped PY2 CI?! Does it mean this PR is bug-free now? laughing

Oops! I originally added @niboshi instead you, but now I've corrected that. Can you see the invite?

**Comment by niboshi** [comment]
>@niboshi we just dropped PY2 CI?! Does it mean this PR is bug-free now? 😆 

Ah, yes that's correct.

We need as well to fix Jenkins configuration to skip Python 2, though.
https://github.com/chainer/chainer/issues/8005

File: tests/cupy_tests/core_tests/test_raw.py [hunk 4]
```diff
@@ -86,6 +97,15 @@ def setUp(self):
         self.mod2 = cupy.RawModule(_test_source2)
         self.mod3 = cupy.RawModule(_test_source3, ("-DPRECISION=2",))
 
+    def tearDown(self):
+        # To avoid cache interference, we remove cached files after every test
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_raw.py]: A test should not clear the kernel cache: it affects too much.
How about temporarily changing the cache directory during the test?
In order for that, we will need a function to set the cache directory (can be private (prefix with `_`) for now).

**Comment by niboshi** [comment]
@leofang 
Thanks. Can you clean up using `try... finlly...` clause?

Also, you can't always expect a change to an environment variable is applied to the current process. You will need a direct way to set the cache directory.

**Comment by leofang** [comment]
Hi @niboshi, I don't think I fully understand you replies...Could you please clarify for me:
> Can you clean up using try... finally... clause? 

Did you refer to the latest commit 35874a2, or the code segment in the `TemporaryDirectory` class that I quoted? Note that the latter isn't part of this PR; it's already there 4 years ago (and should really have been replaced entirely by modern Python's `tempfile.TemporaryDirectory`).

If you referred to the former, did you mean applying the try-finally clause within each and every test? I can certainly do that, but from the viewpoint of writing pythonic test suite wouldn't `setUp()`/`tearDown` be better? It avoids code repetition for sure. 

> Also, you can't always expect a change to an environment variable to be applied to the current process. You will need a direct way to set the cache directory.

Same question as above -- which part were you referring to? 

For the current tests in 35874a2:
1. it's necessary to overwrite `CUPY_CACHE_DIR` in order to let the temp dir propagate all the way to the `cupy.cuda.compiler` module, otherwise it's not possible to overwrite the cache position
2. `os.environ` is accessed effectively as a global  dict, so any update of it is immediately visible within the same process
3. the cache directory is randomly generated using `tempfile.mkdtemp()`, since I need to control the cleanup within a loop in the same test. I can always split the loop into separate tests to circumvent complication. 

Thanks.

**Comment by niboshi** [comment]
Sorry, my coment was ambiguous. I'll answer the questions as follows.

>Did you refer to the latest commit 35874a2, or the code segment in the `TemporaryDirectory` class that I quoted?

I referred the former, and

>If you referred to the former, did you mean applying the try-finally clause within each and every test? I can certainly do that, but from the viewpoint of writing pythonic test suite wouldn't `setUp()`/`tearDown` be better? It avoids code repetition for sure. 

I agree that this kind of cleanup should better be written in `tearDown`. 
What I thought you could improve is this internal loop:

https://github.com/cupy/cupy/blob/35874a25c853aef5c313bcd72a78b175dfe9f587/tests/cupy_tests/core_tests/test_raw.py#L168-L175

This could be written in this manner, which is more pythonic:
```py
for backend in ("nvrtc", "nvcc"):
    try:
        self._setup_test_dir()
        ...
    finally:
        self._remove_test_dir()
```
It's further more pythonic to write this setup-cleanup pair using `setUp`/`tearDown` as you mentioned, but in order for that, the test should first refactored so that "nvrtc" and "nvcc" patterns are separate tests, which is more demanding.

As for the second part,

>2. `os.environ` is accessed effectively as a global  dict, so any update of it is immediately visible within the same process

I think this statement is not actually true.
According to [the documentation](https://docs.python.org/3/library/os.html#os.environ) of `os.environ`:

>If the platform supports the putenv() function, this mapping may be used to modify the environment as well as query the environment. putenv() will be called automatically when the mapping is modified.

>Note: On some platforms, including FreeBSD and Mac OS X, setting environ may cause memory leaks. Refer to the system documentation for putenv(). 

>If the platform supports the unsetenv() function, you can delete items in this mapping to unset environment variables. unsetenv() will be called automatically when an item is deleted from os.environ, and when one of the pop() or clear() methods is called.

So I assume this usage of `os.environ` is not portable across platforms.
But after that I found there are several tests in CuPy that relies on this usage.
That means this issue is not specific to this PR: I think we can continue with this code for now.

**Comment by niboshi** [comment]
https://github.com/cupy/cupy/pull/1941#issuecomment-525809124
As for this comment:

I agree the temporary directory should be removed even if an error happens, but as it's a temorary directory (remnants are meant to be cleaned up regularly), I think it doesn't matter that much.

**Comment by leofang** [comment]
> It's further more pythonic to write this setup-cleanup pair using `setUp`/`tearDown` as you mentioned, but in order for that, the test should first refactored so that "nvrtc" and "nvcc" patterns are separate tests, which is more demanding.

I'm convinced that this is the better way to go, as it'd be important for all tests to run on both backends. In commit 457fba8 I just did this.  

> As for the second part,
> 
> > `os.environ` is accessed effectively as a global  dict, so any update of it is immediately visible within the same process
> 
> I think this statement is not actually true.
> According to [the documentation](https://docs.python.org/3/library/os.html#os.environ) of `os.environ`:

It is true in the strict sense that subsequent CuPy maneuvers in the same test process will see the change in `CUPY_CACHE_DIR`. Other concerns are irrelevant here I think. 

> So I assume this usage of `os.environ` is not portable across platforms.
> But after that I found there are several tests in CuPy that relies on this usage.
> That means this issue is not specific to this PR: I think we can continue with this code for now.

OK, thanks!

This is a little digression now: I am quite surprised to learn about the memory leak issue in `os.environ` -- thanks for pointing that out. But not just CuPy, there're plenty of projects making changes to `os.environ` routinely, especially when using `multiprocessing`. I tend to consider it is the upstream (that is, Python) maintainers' responsibility to make it safe, say by using `setenv()`. Take a step back, I never heard of any issues manipulating  `os.environ`, so perhaps it's not as unsafe in practice.

**Comment by leofang** [comment]
Dear @niboshi, does the latest change c3f8db6 with parameterized tests look better to you? If it's OK, could you kick off Jenkins for me? Thanks!

**Comment by niboshi** [comment]
OK, thank you for the change.
Jenkins, test this please

**Comment by leofang** [comment]
The only failure is again in PY2. (I thought it's retired?) Should I spend time on addressing it? I suspect it's just a matter of incorrect unicode encoding/decoding, but I am not 100% sure. Have to set up a PY2 environment to investigate...

**Comment by niboshi** [comment]
We don't have to handle that.
We need to wait for https://github.com/chainer/chainer-test/pull/523 to get merged.

**Comment by leofang** [comment]
In 54ad318 I fixed `TypeError: Expected unicode, got str` observed in Jenkins. But I was unable to reproduce the other error (`CUDADriverError: CUDA_ERROR_INVALID_SOURCE: device kernel image is invalid`)...Looks like the latter happened when `nvcc` is used, but everything worked fine locally.

**Comment by leofang** [comment]
Hi @niboshi could you kick off Jenkins one more time for me please? In d4ff126 I resolved the conflict in `compiler.py` with #1094, which just went into master. I expect to see only one failure due to PY2. Thanks.

**Comment by niboshi** [comment]
OK.
Jenkins, test this please

**Comment by leofang** [comment]
@niboshi Looks good, thanks. This patch should fix the two build test errors caused by #1094:
```diff
diff --git a/tests/install_tests/test_build.py b/tests/install_tests/test_build.py
index d3fb77c..f90685b 100644
--- a/tests/install_tests/test_build.py
+++ b/tests/install_tests/test_build.py
@@ -12,7 +12,7 @@ class TestCheckVersion(unittest.TestCase):
     def setUp(self):
         self.compiler = ccompiler.new_compiler()
         sysconfig.customize_compiler(self.compiler)
-        self.settings = build.get_compiler_setting()
+        self.settings = build.get_compiler_setting(False)
 
     @pytest.mark.gpu
     def test_check_cuda_version(self):
```
Would you like me to create a PR, or you're already fixing it?

**Comment by leofang** [comment]
> In [54ad318](https://github.com/cupy/cupy/commit/54ad3187fb4c7170c8e25c0a47d538dc4d56521a) I fixed `TypeError: Expected unicode, got str` observed in Jenkins. 

Jenkins confirmed this fix btw.

**Comment by leofang** [comment]
@niboshi Out of curiosity, I was revisiting the failure in PY2. I still feel it's due to an unnoticed subtle bug...

I noticed that the current `nvcc` command missed the `-cubin` option, so I just added it in 8415293. Usually this option is needed, so in this sense I am not sure why PY3 did not fail. Even more weird, my local test in PY2 passed with or without this option.

Anyway, I wonder if we can give it one more try. If PY3 fails  I'll revert the change by force-pushing. Thanks.

**Comment by leofang** [comment]
@niboshi I did three things:
1. First revert back to d4ff126 as the `-cubin` flag was already added
2. Enforce the correct backend selection in the tests
3. Clear the preprocessing cache in `tearDown()`

 I suspect there're some subtle cubin shadowing in the PY2 failure. Not sure if the above changes would fix it, but they're necessary anyway. 

Could you kick off Jenkins again for me please? Sorry for keeping bugging you. Thanks.

**Comment by niboshi** [comment]
https://github.com/chainer/chainer-test/pull/523 has been merged.

@leofang Sure, sorry for delay.
Jenkins, test this please

**Comment by leofang** [comment]
Ah, too bad PY2 is disabled! I'd love to know where it was fixed...Anyway, thanks for moving forward @niboshi.

File: cupy/cuda/compiler.py [hunk 5]
```diff
@@ -174,10 +259,17 @@ def _compile_with_cache_nvrtc(source, options, arch, cache_dir,
...
                 return mod
 
-    ptx = compile_using_nvrtc(source, options, arch, name + '.cu')
-    ls = function.LinkState()
-    ls.add_ptr_data(ptx, u'cupy.ptx')
-    cubin = ls.complete()
+    if backend == 'nvrtc':
+        ptx = compile_using_nvrtc(source, options, arch, name + '.cu')
+        ls = function.LinkState()
+        ls.add_ptr_data(ptx, u'cupy.ptx')
```
Comment (niboshi) [inline cupy/cuda/compiler.py]: In Python3 it's equivalent.

```suggestion
        ls.add_ptr_data(ptx, 'cupy.ptx')
```

File: tests/cupy_tests/core_tests/test_raw.py [hunk 6]
```diff
@@ -1,7 +1,12 @@
+import os
 import pytest
+import shutil
+from tempfile import mkdtemp
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_raw.py]: Please do not import non-module object. (as per OpenStack guideline)

```suggestion
import tempfile
```

File: tests/cupy_tests/core_tests/test_raw.py [hunk 7]
```diff
@@ -78,13 +83,41 @@
...
-        self.mod2 = cupy.RawModule(_test_source2)
-        self.mod3 = cupy.RawModule(_test_source3, ("-DPRECISION=2",))
+        global _test_cache_dir
+        _test_cache_dir = mkdtemp()
+        os.environ['CUPY_CACHE_DIR'] = _test_cache_dir
+
+        self.kern = cupy.RawKernel(_test_source1, 'test_sum',
+                                   backend=self.backend)
+        self.mod2 = cupy.RawModule(_test_source2, backend=self.backend)
+        self.mod3 = cupy.RawModule(_test_source3, ("-DPRECISION=2",),
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_raw.py]: Let's use single quotes for consistency (even though this code has existed from before).

```suggestion
        self.mod3 = cupy.RawModule(_test_source3, ('-DPRECISION=2',),
```

File: tests/cupy_tests/core_tests/test_raw.py [hunk 8]
```diff
@@ -139,16 +172,17 @@ def test_compiler_flag(self):
 
     def test_invalid_compiler_flag(self):
         with pytest.raises(cupy.cuda.compiler.CompileException) as ex:
-            cupy.RawModule(_test_source3, ("-DPRECISION=3",))
+            cupy.RawModule(_test_source3, ("-DPRECISION=3",),
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_raw.py]:
```suggestion
            cupy.RawModule(_test_source3, ('-DPRECISION=3',),
```

File: tests/cupy_tests/core_tests/test_raw.py [hunk 9]
```diff
@@ -139,16 +172,17 @@ def test_compiler_flag(self):
...
         assert 'precision not supported' in str(ex.value)
 
     def test_module_load_failure(self):
         # in principle this test is better done in test_driver.py, but
         # this error is more likely to appear when using RawModule, so
         # let us do it here
-        import os
         with pytest.raises(cupy.cuda.driver.CUDADriverError) as ex:
-            cupy.RawModule(os.path.expanduser("~/this_does_not_exist.cubin"))
+            cupy.RawModule(os.path.expanduser("~/this_does_not_exist.cubin"),
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_raw.py]:
```suggestion
            cupy.RawModule(os.path.expanduser('~/this_does_not_exist.cubin'),
```

File: tests/cupy_tests/core_tests/test_raw.py [hunk 10]
```diff
@@ -78,13 +83,41 @@
...
+    'backend': ('nvrtc', 'nvcc'),
+}))
 class TestRaw(unittest.TestCase):
 
     def setUp(self):
-        self.kern = cupy.RawKernel(_test_source1, 'test_sum')
-        self.mod2 = cupy.RawModule(_test_source2)
-        self.mod3 = cupy.RawModule(_test_source3, ("-DPRECISION=2",))
+        global _test_cache_dir
+        _test_cache_dir = mkdtemp()
```
Comment (leofang) [inline tests/cupy_tests/core_tests/test_raw.py]:
```suggestion
        _test_cache_dir = tempfile.mkdtemp()
```

**Comment by leofang** [comment]
Thank you, @niboshi. Just committed.

**Comment by sjperkins** [comment]
@leofang Thanks for taking over the PR!

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Thanks to @sjperkins for laying the groundwork and to @niboshi for your help and patience!

**Pull request merged**